src/garage/torch/algos/sac.py

```diff
@@ -0,0 +1,258 @@
+
+from garage.np.algos.off_policy_rl_algorithm import OffPolicyRLAlgorithm
+from garage.torch.utils import np_to_torch, torch_to_np
+
+
+class SAC(OffPolicyRLAlgorithm):
+    """ A SAC Model in Torch.
+
+    Soft Actor Critic (SAC) is an algorithm which optimizes a stochastic
+    policy in an off-policy way, forming a bridge between stochastic policy
+    optimization and DDPG-style approaches.
+    A central feature of SAC is entropy regularization. The policy is trained
+    to maximize a trade-off between expected return and entropy, a measure of
+    randomness in the policy. This has a close connection to the
+    exploration-exploitation trade-off: increasing entropy results in more
+    exploration, which can accelerate learning later on. It can also prevent
+    the policy from prematurely converging to a bad local optimum.
+    """
+
+    def __init__(self,
+                 env_spec,
+                 policy,
+                 qf1,
+                 qf2,
+                 target_qf1,
+                 target_qf2,
+                 replay_buffer,
+                 alpha=None,
+                 gradient_steps_per_itr=1000,
+                 target_entropy=None,
+                 initial_log_entropy=0.,
+                 use_automatic_entropy_tuning=False,
+                 discount=0.99,
+                 max_path_length=None,
+                 buffer_batch_size=64,
+                 min_buffer_size=int(1e4),
+                 target_update_tau=5e-3,
+                 policy_lr=3e-4,
+                 qf_lr=3e-4,
+                 optimizer=torch.optim.Adam,
+                 clip_pos_returns=False,
+                 clip_return=np.inf,
+                 max_action=None,
+                 smooth_return=True,
+                 input_include_goal=False):
+
+        self.policy = policy
+        self.qf1 = qf1
+        self.qf2 = qf2
+        self.replay_buffer = replay_buffer
+        self.tau = target_update_tau
+        self.policy_lr = policy_lr
+        self.qf_lr = qf_lr
+        self.initial_log_entropy = initial_log_entropy
+        self.gradient_steps = gradient_steps_per_itr
+        self.clip_pos_returns = clip_pos_returns
+        self.clip_return = clip_return
+        self.evaluate = False
+        self.input_include_goal = input_include_goal
+
+        super().__init__(env_spec=env_spec,
+                         policy=policy,
+                         qf=qf1,
+                         n_train_steps=self.gradient_steps,
+                         n_epoch_cycles=1,
+                         max_path_length=max_path_length,
+                         buffer_batch_size=buffer_batch_size,
+                         min_buffer_size=min_buffer_size,
+                         replay_buffer=replay_buffer,
+                         use_target=True,
+                         discount=discount,
+                         smooth_return=smooth_return)
+
+        # use 2 target q networks
+        self.target_qf1 = copy.deepcopy(self.qf1)
```

Comment:
i thought you accept these as arguments???

Reply:
No, there are four total Q functions in SAC. There's the two Q functions, each of which has a copy with low-pass filtered parameters called the "target" Q function.